src/mbgl/tile/vector_tile.cpp

```diff
@@ -84,6 +84,27 @@ optional<Value> VectorTileFeature::getValue(const std::string& key) const {
     return optional<Value>();
 }
 
+std::unordered_map<std::string,Value> VectorTileFeature::getProperties() const {
+    std::unordered_map<std::string,Value> properties;
+    pbf tags = tags_pbf;
+    while (tags) {
+        uint32_t tag_key = tags.varint();
+        uint32_t tag_val = tags.varint();
+        // this shouldn't use a linear searach of a map
```

Comment:
We could make `VectorTileLayer::keys` a `std::vector<std::string>`, and then make an auxiliary `std::map<std::string, uint32_t>` for use where `keys` is currently used. Ideally the latter would be an `unordered_map`, and the keys would be references into the vector rather than copies of the same strings.